File: src/solutions/2023/09/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false

let ( +/ ) = Int64.add
let ( -/ ) = Int64.sub

let difference ary sz =
  for i = 0 to sz - 1 do
    ary.(i) <- ary.(i + 1) -/ ary.(i)
  done;
  ary
;;

let rec extrapolate ary sz =
  let z = sz - 1 in
  let last = ary.(z) in
  if z = 0 then last else extrapolate (difference ary z) z +/ last
;;

let part_1 (input : string) (_is_test : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.map (fun x ->
    x |> String.split_on_char ' ' |> List.map Int64.of_string |> Array.of_list)
  |> List.map (fun x -> extrapolate x (Array.length x))
  |> List.fold_left ( +/ ) 0L
  |> Int64.to_string
;;

let part_2 (input : string) (_is_test : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.map (fun x ->
    x |> String.split_on_char ' ' |> List.map Int64.of_string |> List.rev |> Array.of_list)
  |> List.map (fun x -> extrapolate x (Array.length x))
  |> List.fold_left ( +/ ) 0L
  |> Int64.to_string
;;

let () = Aoc.run Sys.argv part_1 part_2 options

File: src/solutions/2023/09/python/main.py
import sys
import os
from typing import List

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': False, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


def difference(history: List[int], length: int) -> List[int]:
   for i in range(length):
      history[i] = history[i + 1] - history[i]
   return history


def extrapolate(history: List[int], length: int) -> int:
   length = length - 1
   last = history[length]
   if length == 0:
      return last
   return extrapolate(difference(history, length), length) + last


def part_one(input: str, is_test: bool) -> str:
   parsed = [[int(y) for y in x.split(' ')] for x in input.split('\n')]
   res = 0
   for history in parsed:
      res += extrapolate(history, len(history))
   return str(res)


def part_two(input: str, is_test: bool) -> str:
   parsed = [[int(y) for y in x.split(' ')] for x in input.split('\n')]
   res = 0
   for history in parsed:
      history.reverse()
      res += extrapolate(history, len(history))
   return str(res)


if __name__ == '__main__':
   main()
